### DISCUSSION TITLE
Introduce new concept: Offers

### DISCUSSION THREAD

**Discussion by Arminek** [Sylius/Sylius#13180] [Ideas]
# 👋  Hello, Sylius community,
According to our [roadmap](https://sylius.com/roadmap), we have started thinking about tackling these new features. These are kind of related to each other:
- Catalog promotions
- Promotion Campaigns
- Product Drafts
- Pre-Sale Campaigns

### So yeah, we would like to start with stop & check. Let's review what we already have inside Sylius.

From [Archetype Patterns](https://www.amazon.pl/Arlow-Enterprise-Patterns-Building-Archetype/dp/032111230X) perspective in Sylius we have covered (not fully):
- The Product Archetyp - describes the common properties of a set of goods or services. (Sylius Product model)
- The Product Instance Archetype - represents a specific instance of a Product Type. (Sylius Product Variant model)

When we have to go through the new features which we plan to introduce. We have noticed that there is a missing core concept that will allow us to do it in a more elegant way. The missing archetype pattern is called the Product Catalog.

**The Product Catalog archetype** - represents a persistent store of product information used in the selling process.
So we want to introduce a new core concept of offers inside sylius standard. But what does it mean from a technical point of view?

Offers will connect few things together:
- list of variants
- price
- validity period

We are thinking about introducing `OfferSearchEngineInterface` with default implementation based on elastic search (ES in sylius core by default).

We believe that features like: 
- Personalized offers
- Catalog time-based promotions
- Drafting an offer
Will be much easier to implement using this pattern.

This is just an idea. We would love to hear our community 😄 

I want also to invite you to this [board](https://github.com/Sylius/Sylius/projects/8) if you are interested in topics related to the future sylius architecture.

**Comment by pjedrzejewski** [comment]
Hey! Awesome issue @Arminek, thanks!

I have a couple of thoughts/ideas related to this:

1) In the future versions (most likely 3.x) I'd like to get rid of the Product and Product Variant differentiation and introduce a [holarchical](https://en.wikipedia.org/wiki/Holarchy) (hierarchical with a twist) structure of just Product. With the current modeling: Product and Product Variant is a hierarchy already (but with a limit of 1 level), which is a nice start but as we evolved the platform, both models are very similar now. Few examples of how it could look:

Physical products:

- Product: The Witcher T-Shirt
   - Product: The Witcher T-Shirt (Black)
      - Product: The Witcher T-Shirt (Black/S)
      - Product: The Witcher T-Shirt (Black/M) etc.
    - Product: The Witcher T-Shirt (White)
      - Product: The Witcher T-Shirt (White/S)
      - Product: The Witcher T-Shirt (White/M) etc.

Digital products:

- Product: Sylius Core
   - Product: Sylius Plus
   - Product: BitBagCmsPlugin

To better understand holarchies, you can do a quick mind puzzle: Would a concept of a white Witcher t-shirt with size S exist without a concept of a Witcher t-shirt product at all and of course the white variation of it? Same goes for digital products, BitBagCmsPlugin would not exist if Sylius Core does not exist etc. It's an useful check for verifying holarchies. Each holon (lower) also transcends (goes beyond/adds more depth or value, you name it) and includes the other (higher) holon. It's true for both t-shirts and Sylius', etc. Another interesting example would be a novel, which itself is a Product but its softback copy, hardhack copy "anniversary edition" and e-book version for Kindle are also Products. You can check here that the novel itself has some VALUE but a different version of it, go beyond that value and add something "on top". I am yet to find something of value that would not fit this model but I am very keen to have it challenged. :)

Currently, it could be cumbersome but if we introduce a new model "Offering", which would be a combination of product, price, timeframe and maybe other rules (like a customer segment/group, specific market, etc.), it would be easier, cause only Products which have some Offering are sellable/exchangable, while others are just **intellectual property**.

2) ^ this would also enable few other cool things (not without challenges, of course :D), like: removing the differentiation between Attributes and Options, they are a duplication to some degree right now. I think Akeneo has done it quite well, we can take some inspiration. :) Another thing would be the attributes inheritance. So any attributes common to the entire holarchy of products,  could be defined only on the highest relevant element and below that, only specific attributes (like color, size, weight) would differ.

3) I think we have a different perspective on what is the Product Instance. In my view it's not a Product Variant (see reasoning above) but a specific instance of a Product. Let's take some examples...

For a physical product, it would be a specific The Witcher T-Shirt (Black/M) that is laying in the Lyon Warehouse, section B, shelf 48, position 5. Or a specific t-shirt we shipped to John Doe in Boston that is now going back to us cause he is doing a return. And it's different from another Product Instance of a the same t-shirt, but size S that he has ordered for a friend and it's ok.

For a digital product, well a Product Instance could be hbx.com (an instance of Product: Sylius Core) or bakken.nl (instance of Product: Sylius Plus). I was plying with the idea of tying it to licensing but it's actually different, cause you can: 

1. create a Product: ERP Integration App for Sylius
2. offer (Offering, wink wink) a License for this product
3. allow up to 3 or unlimited Product Instances as part of these particular licensing terms
4. whoever buys the license, can spin up X instances of that product.

So long story, short if we want to deal with intellectual property and licensing properly, we would need an additional model IMO.

What are your thoughts on this? (PS. Yes, I will get the book, one day!)

4) I was also experimenting with a common model for Products and Services and I am on the edge here, I think we need to test it in practice at Sylius internally. We do not want to design these models but discover them! They already exist, we just need to be patience, humble and 'have a good eye'! 8) Let's take an example of the most common service performed by Sylius-powered businesses right now: the delivery.

As a merchant, I offer you some products and if they are physical, I offer a Service: Shipping, the holarchy (yes!) could look something like this:

- Service: Delivery
   - DHL
      - DHL Express (maybe it has some parameters, like preferred time)
   - InPost Parcel Lockers (maybe it has some parameters too, like the specific locker location)

All can be performed, including the generic "Delivery", cause if you care about the customer's experience, if all shipping providers fail, you can rent a bike or helicopter and still make it. In the end, the customer just wants to get it. :)

Or our (Sylius) consulting services:

- Service: Sylius Consulting Services
   - Service: Architecture Workshop
   - Service: General Audit
       -  Service: Performance Audit

Each adds more depth (in that case detail and focus), all can be offered, etc.

So, in theory, we could squeeze it into Product model but something is odd to me about this idea, not sure what yet because I see a lot of similarities. That being said, in the case of a service, we talk more about specific "Service Delivery" (like in consulting/development delivery process, etc.). Also, the other value ingredient that I see (except Product, Service and Licence) is Relationship but I leave that for another long post. :))) We could consider extracting something common out of it but let's make it an exciting discovery through experimentation process rather than guessing!

**Comment by loevgaard** [comment]
I think for people to be able to grasp this (at least I am talking for myself here) and thereby contribute here, it would be beneficial to add some scenarios where this feature will make so much sense

**Comment by plewandowski** [comment]
Well if it is time to re-think a product then what about addressing products with decimal quantities and different units?

**Comment by XaBe20** [comment]
Hey Sylius community! 

Very good idea @pjedrzejewski to have a holarchical structure for products with many levels 👍 

From my side too I have several ideas:
- Decreasing price management according to quantity without promotion rule but with price list management like this : 1 product 10€ / 2 products 18€ (9€/piece) / 5 products 25€ (5€/piece)....
- Quote management that can be converted into an order : from the front shopping cart view possibility of creating a quote, then in back-office to be able to validate the quote so that the user can convert it to the front

**Comment by 4c0n** [comment]
I would like to point out YAGNI: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

Whereas the global consensus is that you shouldn't be basing decisions you need to make right now on things that you think you might need in the future. So if the discussion is currently about adding a way to ease the use of "offers", we shouldn't be talking about all sorts of other things. Instead when it gets to that point you can decide to rework what you have in a new major version so there's no need to be scared of BC breaks etc.

Instead of complicating things, it might be better to first define exactly the use cases you want to address and take it from there (like @loevgaard pointed out). Then maybe the patterns will make more sense. Right now it just kind of seems unclear what exactly the purpose is of adding this additional complexity, I'd say this goes for both ideas.

Also instead of referring to a book on the Polish Amazon shop, it would be better to refer to for example wikipedia: https://en.wikipedia.org/wiki/Archetype_pattern so that if people are unaware of the pattern they can quickly get a grasp of what it means without having to buy and read a book about it.

Personally I think there are bigger issues to solve first, but that's more of a timeline thing.

**Reply by Arminek** [reply]
@4c0n Thank you for pointing this out. There are some features that we are going to address in near future. Next Q1 ‘22 – Q3 ‘22 & Future (Ideas) according to [the roadmap](https://sylius.com/roadmap/). And this is just a discussion of how we **could** address these needs.

**Reply by 4c0n** [reply]
@Arminek The goal of the discussion is quite clear to me, but thanks for repeating it once more.

To clarify even more, there's a suggestion of doing it a certain way, because maybe in 2 major versions from the current one, some ideas might be implemented (conditional on experimentation which may or may not amount to something necessary/useful), also there is the suggestion of integrating elastic search support in the core for example. I'd say YAGNI applies very well in those cases, do you disagree? If so it would benefit the discussion if you would tell us why instead of just giving the comment a thumbs down.

You must understand that in a discussion not everyone is just going to agree right? It would be more of an announcement that way.

**Reply by Arminek** [reply]
@4c0n Why do you have to be mad xD? I do not like your response. I do not want everyone to agree, definitely no. But I want to talk about ideas on how to tackle some things. Here is a counterargument for your [YAGNI](https://rocket-science.ru/tips/2018/03/06/beware-of-yagni).

**Reply by 4c0n** [reply]
Who says I'm mad about anything? Just pointing out some well known paradigm, that you happen to disagree with and the writer of the blog does not agree with either. I'm sure there are also blogs that state the opposite, however they may or may not apply to this specific case, so I'm sorry but it's not the strongest argument IMO.

My sincere apologies that you find my opinion to be "just moaning". I guess everyone lives by their own truths.

I hope you will get more relevant and useful input in the rest of the discussion. Which was kind of what I was trying to achieve with my original comment. 😅

**Reply by Arminek** [reply]
What I get from this discussion is that there is a need for a better explanation:

- User stories to showcase the problem we want to solve. (Not just a proposition of the solution from the thin air.)
- How could it work? (visualizations)
- What does it mean in terms of current concepts. Do we plan BC etc.? (We do not. [BC Promise](https://docs.sylius.com/en/1.10/book/organization/backward-compatibility-promise.html))

To conclude: this discussion was overall useful. Thanks for bringing the topic. Nonetheless, I do not appreciate fully how certain concepts were brought up to light.

**Comment by Arminek** [comment]
I will prepare some stories that will show examples of using this concept & the problem that we want to solve. Stay tuned 💃

**Comment by nikkuexe** [comment]
Is there some thought on using [product types (with attribute sets)](https://github.com/Sylius/Sylius/issues/9678) as a way to make it easier to create offers? Here are some cases:

A discount on all pants in the color black or orange (but not shirts in the color black, or shoes in the color orange)
A discount on all laptops with AMD CPUs and AMD GPUs on sale (but not laptops with a Intel CPU and AMD GPU)

**Comment by gabiudrescu** [comment]
Guys, i want to take a simple approach on this one: what problem are you trying to solve with this?🤔

**Comment by dbalabka** [comment]
> We are thinking about introducing OfferSearchEngineInterface with default implementation based on elastic search (ES in sylius core by default).

>Currently, it could be cumbersome but if we introduce a new model "Offering", which would be a combination of product, price, timeframe and maybe other rules (like a customer segment/group, specific market, etc.), it would be easier, cause only Products which have some Offering are sellable/exchangable, while others are just intellectual property.

@pjedrzejewski, it is a great step forward that you are going to include ES support into Sylius by default. I have some experience with implementing different strategies of pricing using ES. Could you please put light on how you have going to deal with multiple different prices for the products and still support complex ES aggregation (e.g., nested aggregation) and sorting of prices? Also, it would be interesting how you are going to keep multiple prices support on MySQL.

Could you please share your vision about ES and MySQL schema structure?
